File: src/neetcode/Java/_13RomantoInteger.java
package neetcode.Java;

import java.util.HashMap;

public class _13RomantoInteger {
    public static int romanToInt1(String s) {
        String[] val = {"I", "IV", "V", "IX", "X",
                "XL", "L", "XC", "C", "CD", "D", "CM", "M",
        };
        int[] key = {1, 4, 5, 9, 10, 40,
                50, 90, 100, 400, 500, 900, 1000
        };
        int pos = 0;
        int len = s.length();
        int num = 0;
        for (int i = key.length - 1; i >= 0 && pos < len; ) {
            if (pos + 1 < len && (s.charAt(pos) + "" + s.charAt(pos + 1) + "").equals(val[i])) {
                pos += 2;
                num = num + key[i];
                continue;
            }
            if ((s.charAt(pos) + "").equals(val[i])) {
                pos++;
                num = num + key[i];
                continue;
            }
            i--;
        }
        return num;
    }

    public int romanToInt2(String s) {
        HashMap<Character, Integer> h1 = new HashMap<>();
        h1.put('I', 1);
        h1.put('V', 5);
        h1.put('X', 10);
        h1.put('L', 50);
        h1.put('C', 100);
        h1.put('D', 500);
        h1.put('M', 1000);
        char[] alpha = s.toCharArray();
        int num = 0;
        int pt = 0;
        while (pt + 1 <= alpha.length) {
            if (pt == alpha.length - 1 || h1.get(alpha[pt]) >= h1.get(alpha[pt + 1])) {
                num = num + h1.get(alpha[pt]);
                pt = pt + 1;
            } else {
                num = num + h1.get(alpha[pt + 1]) - h1.get(alpha[pt]);
                pt = pt + 2;
            }
        }
        return num;
    }
        public static int romanToInt (String s){
            String[] val = {"I", "IV", "V", "IX", "X",
                    "XL", "L", "XC", "C", "CD", "D", "CM", "M",
            };
            int[] key = {1, 4, 5, 9, 10, 40,
                    50, 90, 100, 400, 500, 900, 1000
            };
            HashMap<String, Integer> h1 = new HashMap();
            h1.put("I", 1);
            h1.put("IV", 4);
            h1.put("V", 5);
            h1.put("IX", 9);
            h1.put("X", 10);
            h1.put("XL", 40);
            h1.put("L", 50);
            h1.put("XC", 90);
            h1.put("C", 100);
            h1.put("CD", 400);
            h1.put("D", 500);
            h1.put("CM", 900);
            h1.put("M", 1000);
            char[] ss = s.toCharArray();
            int pos = 0;
            int len = s.length();
            int num = 0;
            for (int i = key.length - 1; i >= 0 && pos < len; ) {
                if (pos + 1 < len && (ss[pos] + "" + ss[pos + 1]).equals(val[i])) {
                    pos += 2;
                    num = num + key[i];
                    continue;
                }
                if ((ss[pos] + "").equals(val[i])) {
                    pos++;
                    num = num + key[i];
                    continue;
                }
                i--;
            }
            return num;
        }

        public static void main (String[]args){
            int res = romanToInt("MCMXCIV");
            System.out.println(res);
        }
    }

File: src/neetcode/kotlin/_13RomantoInteger.kt
package neetcode.kotlin

class _13RomantoInteger {
    fun romanToInt2(s: String): Int {
        val h1 = HashMap<Char, Int>()
        h1['I'] = 1
        h1['V'] = 5
        h1['X'] = 10
        h1['L'] = 50
        h1['C'] = 100
        h1['D'] = 500
        h1['M'] = 1000
        val alpha = s.toCharArray()
        var num = 0
        var pt = 0
        while (pt + 1 <= alpha.size) {
            if (pt == alpha.size - 1 || h1[alpha[pt]]!! >= h1[alpha[pt + 1]]!!) {
                num = num + h1[alpha[pt]]!!
                pt = pt + 1
            } else {
                num = num + h1[alpha[pt + 1]]!! - h1[alpha[pt]]!!
                pt = pt + 2
            }
        }
        return num
    }

    companion object {
        fun romanToInt1(s: String): Int {
            val `val` = arrayOf(
                "I", "IV", "V", "IX", "X",
                "XL", "L", "XC", "C", "CD", "D", "CM", "M",
            )
            val key = intArrayOf(
                1, 4, 5, 9, 10, 40,
                50, 90, 100, 400, 500, 900, 1000
            )
            var pos = 0
            val len = s.length
            var num = 0
            var i = key.size - 1
            while (i >= 0 && pos < len) {
                if (pos + 1 < len && (s[pos].toString() + "" + s[pos + 1] + "") == `val`[i]) {
                    pos += 2
                    num = num + key[i]
                    continue
                }
                if ((s[pos].toString() + "") == `val`[i]) {
                    pos++
                    num = num + key[i]
                    continue
                }
                i--
            }
            return num
        }

        fun romanToInt(s: String): Int {
            val `val` = arrayOf(
                "I", "IV", "V", "IX", "X",
                "XL", "L", "XC", "C", "CD", "D", "CM", "M",
            )
            val key = intArrayOf(
                1, 4, 5, 9, 10, 40,
                50, 90, 100, 400, 500, 900, 1000
            )
            val h1: HashMap<Any?, Any?> = HashMap<Any?, Any?>()
            h1["I"] = 1
            h1["IV"] = 4
            h1["V"] = 5
            h1["IX"] = 9
            h1["X"] = 10
            h1["XL"] = 40
            h1["L"] = 50
            h1["XC"] = 90
            h1["C"] = 100
            h1["CD"] = 400
            h1["D"] = 500
            h1["CM"] = 900
            h1["M"] = 1000
            val ss = s.toCharArray()
            var pos = 0
            val len = s.length
            var num = 0
            var i = key.size - 1
            while (i >= 0 && pos < len) {
                if (pos + 1 < len && (ss[pos].toString() + "" + ss[pos + 1]) == `val`[i]) {
                    pos += 2
                    num = num + key[i]
                    continue
                }
                if ((ss[pos].toString() + "") == `val`[i]) {
                    pos++
                    num = num + key[i]
                    continue
                }
                i--
            }
            return num
        }

        @JvmStatic
        fun main(args: Array<String>) {
            val res = romanToInt("MCMXCIV")
            println(res)
        }
    }
}
